### ISSUE TITLE
distributed-key-value-store example fails to PUT, rejected protocol /ipfs/kad/1.0.0

### ISSUE THREAD

**Issue by 0x7CFE** [libp2p/rust-libp2p#4194]
Revision 4c889fd6d5eface44aca38a8fa1664b7248c62ca

On the first node where I do `PUT hello world`:

```
[2023-07-13T08:42:56Z DEBUG libp2p_swarm] Connection established: PeerId("12D3KooWBebdYFBEJvpjvRPjPvymUZaN2mEt15nzcatpB1ZPqvZt") Dialer { address: "/ip4/172.17.0.1/tcp/44217/p2p/12D3KooWBebdYFBEJvpjvRPjPvymUZaN2mEt15nzcatpB1ZPqvZt", role_override: Dialer }; Total (peer): 1.
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] reading incoming messages
[2023-07-13T08:42:56Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] handling requests
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] sending messages
[2023-07-13T08:42:56Z DEBUG yamux::connection] 66bd1dd8: new outbound (Stream 66bd1dd8/1) of (Connection 66bd1dd8 Client (streams 0))
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] reading incoming messages
[2023-07-13T08:42:56Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] handling requests
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] sending messages
[2023-07-13T08:42:56Z DEBUG multistream_select::dialer_select] Dialer: Proposed protocol: /ipfs/kad/1.0.0
[2023-07-13T08:42:56Z DEBUG multistream_select::negotiated] Negotiated: Received confirmation for protocol: /yamux/1.0.0
[2023-07-13T08:42:56Z DEBUG multistream_select::dialer_select] Dialer: Received rejection of protocol: /ipfs/kad/1.0.0
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] reading incoming messages
[2023-07-13T08:42:56Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] handling requests
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] sending messages
[2023-07-13T08:42:56Z DEBUG libp2p_kad::behaviour] Request to PeerId("12D3KooWBebdYFBEJvpjvRPjPvymUZaN2mEt15nzcatpB1ZPqvZt") in query QueryId(0) failed with Upgrade(NegotiationFailed)
Failed to put record: QuorumFailed { key: Key(b"hello"), success: [], quorum: 1 }
```

On the second node:
```
[2023-07-13T08:42:56Z DEBUG libp2p_swarm] Connection established: PeerId("12D3KooWC7hvYMsgXVbH4q27QnJysxZjGqdvm3W4tjJEmyEwfM43") Listener { local_addr: "/ip4/172.17.0.1/tcp/44217", send_back_addr: "/ip4/192.168.31.180/tcp/36418" }; Total (peer): 1.
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] reading incoming messages
[2023-07-13T08:42:56Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] handling requests
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] sending messages
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] reading incoming messages
[2023-07-13T08:42:56Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] handling requests
[2023-07-13T08:42:56Z DEBUG netlink_proto::connection] sending messages
[2023-07-13T08:42:56Z DEBUG multistream_select::listener_select] Listener: rejecting protocol: /ipfs/kad/1.0.0
[2023-07-13T08:42:58Z DEBUG netlink_proto::connection] reading incoming messages
[2023-07-13T08:42:58Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-07-13T08:42:58Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-07-13T08:42:58Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-07-13T08:42:58Z DEBUG netlink_proto::connection] handling requests
[2023-07-13T08:42:58Z DEBUG netlink_proto::connection] sending messages
[2023-07-13T08:43:06Z DEBUG yamux::connection] 48f9c4c1: socket eof
[2023-07-13T08:43:06Z DEBUG netlink_proto::connection] reading incoming messages
[2023-07-13T08:43:06Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-07-13T08:43:06Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-07-13T08:43:06Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-07-13T08:43:06Z DEBUG netlink_proto::connection] handling requests
[2023-07-13T08:43:06Z DEBUG netlink_proto::connection] sending messages
[2023-07-13T08:43:06Z DEBUG libp2p_swarm] Connection closed with error IO(Custom { kind: Other, error: Error(Closed) }): Connected { endpoint: Listener { local_addr: "/ip4/172.17.0.1/tcp/44217", send_back_addr: "/ip4/192.168.31.180/tcp/36418" }, peer_id: PeerId("12D3KooWC7hvYMsgXVbH4q27QnJysxZjGqdvm3W4tjJEmyEwfM43") }; Total (peer): 0.
```

**Comment by dariusc93** [comment]
I would have to double check but i think the example needs to be updated to add an external address or to use `Kademlia::set_mode` to set it to `Mode::Server`. 

Can you apply one the following patch and see if it will work

```
diff --git a/examples/distributed-key-value-store/src/main.rs b/examples/distributed-key-value-store/src/main.rs
index 226a438f..52d62036 100644
--- a/examples/distributed-key-value-store/src/main.rs
+++ b/examples/distributed-key-value-store/src/main.rs
@@ -98,6 +98,7 @@ async fn main() -> Result<(), Box<dyn Error>> {
         event = swarm.select_next_some() => match event {
             SwarmEvent::NewListenAddr { address, .. } => {
                 println!("Listening in {address:?}");
+                swarm.add_external_address(address);
             },
             SwarmEvent::Behaviour(MyBehaviourEvent::Mdns(mdns::Event::Discovered(list))) => {
                 for (peer_id, multiaddr) in list {
```

```
diff --git a/examples/distributed-key-value-store/src/main.rs b/examples/distributed-key-value-store/src/main.rs
index 226a438f..e3bfc140 100644
--- a/examples/distributed-key-value-store/src/main.rs
+++ b/examples/distributed-key-value-store/src/main.rs
@@ -23,6 +23,7 @@
 use async_std::io;
 use futures::{prelude::*, select};
 use libp2p::core::upgrade::Version;
+use libp2p::kad::Mode;
 use libp2p::kad::record::store::MemoryStore;
 use libp2p::kad::{
     record::Key, AddProviderOk, GetProvidersOk, GetRecordOk, Kademlia, KademliaEvent, PeerRecord,
@@ -85,6 +86,8 @@ async fn main() -> Result<(), Box<dyn Error>> {
         SwarmBuilder::with_async_std_executor(transport, behaviour, local_peer_id).build()
     };
 
+    swarm.behaviour_mut().kademlia.set_mode(Some(Mode::Server));
+
     // Read full lines from stdin
     let mut stdin = io::BufReader::new(io::stdin()).lines().fuse();
 
```

**Comment by 0x7CFE** [comment]
Indeed, setting mode to Server did the trick. I've prepared PR https://github.com/libp2p/rust-libp2p/pull/4195 with that change.

**Comment by tqwewe** [comment]
I'm also having the same issue running the distributed key value store example.

When I run it in two terminals, and do `PUT foo bar`, I get an error:
```
Failed to put record: QuorumFailed { key: Key(b"foo"), success: [], quorum: 1 }
```

I did notice if I do `cargo run` in both terminals at the same time (one server don't start much after the other), then it works. But having a delay of one second or more between the two terminals starting makes the example no longer work.

---

**Edit**

I've just tried on my Linux PC, and I don't have this issue. Seems like it was specific to Mac.
Additionally, on Linux I get 10 "Listening in ..." logs, whilst on Mac I only got one I believe.
